go/private/actions/compilepkg.bzl

```diff
@@ -194,3 +202,89 @@ def emit_compilepkg(
+        _run_nogo(
+            go,
+            sources = sources,
+            importpath = importpath,
+            importmap = importmap,
+            archives = archives + ([cover_archive] if cover_archive else []),
+            recompile_internal_deps = recompile_internal_deps,
+            cover_mode = cover_mode,
+            cgo_go_srcs = cgo_go_srcs_for_nogo,
+            out_facts = out_facts,
+            out_log = out_nogo_log,
+            out_validation = out_nogo_validation,
+            nogo = nogo,
+        )
+
+def _run_nogo(
+        go,
+        *,
+        sources,
+        importpath,
+        importmap,
+        archives,
+        recompile_internal_deps,
+        cover_mode,
+        cgo_go_srcs,
+        out_facts,
+        out_log,
+        out_validation,
+        nogo):
+    """Runs nogo on Go source files, including those generated by cgo."""
+    inputs = (sources + [nogo, go.package_list] +
+              [archive.data.facts_file for archive in archives if archive.data.facts_file] +
+              [archive.data.export_file for archive in archives] +
+              go.sdk.tools + go.sdk.headers + go.stdlib.libs)
+    outputs = [out_facts, out_log]
+
+    args = go.builder_args(go, "nogo", use_path_mapping = True)
+    args.add_all(sources, before_each = "-src")
+    if cgo_go_srcs:
+        inputs.append(cgo_go_srcs)
+        args.add_all([cgo_go_srcs], before_each = "-src")
+    if cover_mode:
+        args.add("-cover_mode", cover_mode)
+    if recompile_internal_deps:
+        args.add_all(recompile_internal_deps, before_each = "-recompile_internal_deps")
+    args.add_all(archives, before_each = "-arc", map_each = _archive)
+    if importpath:
+        args.add("-importpath", importpath)
+    else:
+        args.add("-importpath", go.label.name)
+    if importmap:
+        args.add("-p", importmap)
+    args.add("-package_list", go.package_list)
+
+    args.add_all(archives, before_each = "-facts", map_each = _facts)
+    args.add("-out_facts", out_facts)
+    args.add("-out_log", out_log)
+    args.add("-nogo", nogo)
+
+    go.actions.run(
+        inputs = inputs,
+        outputs = outputs,
+        mnemonic = "RunNogo",
+        executable = go.toolchain._builder,
+        arguments = [args],
+        env = go.env_for_path_mapping,
+        toolchain = GO_TOOLCHAIN_LABEL,
+        execution_requirements = SUPPORTS_PATH_MAPPING_REQUIREMENT,
+        progress_message = "Running nogo on %{label}",
+    )
+
+    validation_args = go.actions.args()
+    validation_args.add("nogovalidation")
+    validation_args.add(out_validation)
+    validation_args.add(out_log)
+    go.actions.run(
+        inputs = [out_log],
+        outputs = [out_validation],
+        mnemonic = "ValidateNogo",
+        executable = go.toolchain._builder,
+        arguments = [validation_args],
+        execution_requirements = SUPPORTS_PATH_MAPPING_REQUIREMENT,
+        progress_message = "Validating nogo output for %{label}",
+    )
```

Comment:
Could you please add a comment here to clarify why we use 2 instead of 1 action? And if there is a "failing" analysis, then which of the 2 actions will fail?

It's slightly confusing to me because if we were to run this with `--sandbox_debug`, then the sandbox would not contain the failing scanned ".go" source file, but only the output log from `RunNogo`. This could make troubleshooting nogo failure a bit more difficult.

Reply:
I added comments to both actions that should address your concerns: `--sandbox_debug` debugging is still possible.

Reply:
So there are 2 failure cases:

- Rules/analyzer author: the analyzer failed to run and die after the first action. This can be sandbox debugged.

- Typical developer: they changed some Go source code, did build, and failed the "linter". This check happens in the second action and currently, the sandbox for this would only include the log file. If user try to debug with sandbox debug, they will only be able to see the log file and nothing else.

What do you think if we reuse all the inputs of the first action in the second action?
I think that would enable users to use sandbox debug to inspect the `.go` source files that are failing the style check.

Reply:
Would you consider it a common case to `--sandbox_debug` a linter failure? Wouldn't the output of `--verbose_failures` or `-s` be sufficient to debug this? `--spawn_strategy=local` is also always an option when you aren't interested in the sandboxing aspect.

I'm hesitant to add unnecessary inputs to an action that can easily lie on the critical path, but it's true that it is very fast and reusing inputs shouldn't be too costly.

Reply:
In the past, we have had nogo running on all sorts of unexpected source files: external deps, coverage generated, cgo generated, etc... So it was common for me to use `--sandbox_debug` to check what was inside the GoCompilePkg sandbox: inputs and leftover outputs.

As nogo improved over time, this has happened less frequently. Still, I would prefer to keep the ability to debug the action and to let users access what caused the failures vs only being able to access the action that processed the failure log.